blocks/graph/bn.py

```diff
@@ -0,0 +1,163 @@
+transformation of a batch-normalized inference graph into training graph,
+which uses minibatch statistics in place of population statistics.
+
+"""
+import collections
+import contextlib
+
+from ..roles import BATCH_NORM_OFFSET, BATCH_NORM_DIVISOR, INPUT, OUTPUT
+from ..utils import find_bricks
+
+
+@contextlib.contextmanager
+def batch_normalization(*bricks):
+    r"""Context manager to run batch normalization in "training mode".
+
+    Parameters
+    ----------
+    \*bricks
+        One or more bricks which will be inspected for descendant
+        instances of :class:`~blocks.bricks.BatchNormalization`.
+
+    Notes
+    -----
+    Graph replacement using :func:`apply_batch_normalization`, while
+    elegant, can lead to Theano graphs that are quite large and result
+    in very slow compiles. This provides an alternative mechanism for
+    building the batch normalized training graph. It can be somewhat
+    less convenient as it requires building the graph twice if one
+    wishes to monitor the output of the inference graph during training.
+
+    Examples
+    --------
+    First, we'll create a :class:`~blocks.bricks.BatchNormalizedMLP`.
+    This behaves almost exactly like a regular :class:`~blocks.bricks.MLP`
+    except that it contains :class:`~blocks.bricks.BatchNormalization`
+    bricks placed before each activation function.
+
+    >>> import theano
+    >>> from blocks.bricks import BatchNormalizedMLP, Tanh
+    >>> from blocks.initialization import Constant, IsotropicGaussian
+    >>> mlp = BatchNormalizedMLP([Tanh(), Tanh()], [4, 5, 6],
+    ...                          weights_init=IsotropicGaussian(0.1),
+    ...                          biases_init=Constant(0))
+    >>> mlp.initialize()
+    >>> x = theano.tensor.matrix('x')
+
+    First, we'll construct an output variable as we would normally. This
+    is getting normalized by the *population* statistics, which by
+    default are initialized to 0 (mean) and 1 (standard deviation),
+    respectively.
+
+    >>> y = mlp.apply(x)
+
+    And now, to construct an output with batch normalization enabled,
+    i.e. normalizing pre-activations using per-minibatch statistics, we
+    simply make a similar call inside of a `with` statement:
+
+    >>> with batch_normalization(mlp):
+    ...     y_bn = mlp.apply(x)
+
+    Let's verify that these two graphs behave differently on the
+    same data:
+
+    >>> import numpy
+    >>> data = numpy.arange(12, dtype=theano.config.floatX).reshape(3, 4)
+    >>> inf_y = y.eval({x: data})
+    >>> trn_y = y_bn.eval({x: data})
+    >>> numpy.allclose(inf_y, trn_y)
+    False
+
+    """
+    # Avoid circular imports.
+    from blocks.bricks import BatchNormalization
+
+    bn = find_bricks(bricks, lambda b: isinstance(b, BatchNormalization))
+    # Can't use either nested() (deprecated) nor ExitStack (not available
+    # on Python 2.7). Well, that sucks.
+    for brick in bn:
+        brick.__enter__()
+    yield
```

Comment:
Maybe wrap this in a try/finally to be sure `__exit__` is called. Also the loop below should technically go over the bricks in reverse.